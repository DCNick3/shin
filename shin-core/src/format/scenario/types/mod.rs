mod list;
mod pad4;

pub use list::{
    SmallList, U16SmallList, U16SmallNumberList, U8SmallList, U8SmallNumberList, SMALL_LIST_SIZE,
};
pub use pad4::Pad4;

pub use crate::format::text::{
    SJisString, StringArray, U16FixupString, U16String, U8FixupString, U8String, ZeroString,
};
