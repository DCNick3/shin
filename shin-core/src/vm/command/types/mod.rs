//! Types used in commands.

mod flags;
mod id;
mod property;

pub use flags::{AudioWaitStatus, LayerCtrlFlags, MaskFlags};
pub use id::{
    LayerId, LayerIdOpt, VLayerId, VLayerIdRepr, LAYERBANKS_COUNT, LAYERS_COUNT, PLANES_COUNT,
};
pub use property::{LayerProperty, LayerPropertySmallList};

use crate::format::scenario::instruction_elements::FromNumber;
use num_derive::FromPrimitive;

#[derive(FromPrimitive, Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum LayerType {
    Null = 0,
    Tile = 1,
    Picture = 2,
    Bustup = 3,
    Animation = 4,
    Effect = 5,
    Movie = 6,
    FocusLine = 7,
    Rain = 8,
    Quiz = 9,
}

impl FromNumber for LayerType {
    fn from_number(number: i32) -> Self {
        num_traits::FromPrimitive::from_i32(number)
            .unwrap_or_else(|| panic!("LayerType::from_vm_ctx: invalid layer type: {}", number))
    }
}

#[derive(FromPrimitive, Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Hash)]
pub enum MessageboxType {
    Neutral = 0,
    WitchSpace = 1,
    Ushiromiya = 2,
    Transparent = 3,
    Novel = 4,
    NoText = 5,
}

#[derive(FromPrimitive, Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Hash)]
pub enum MessageTextLayout {
    Left = 0,
    /// I _think_ this is the same as Left
    Layout1 = 1,
    Center = 2,
    Right = 3,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Hash)]
pub struct MessageboxStyle {
    pub messagebox_type: MessageboxType,
    pub text_layout: MessageTextLayout,
}

impl Default for MessageboxStyle {
    fn default() -> Self {
        Self {
            messagebox_type: MessageboxType::Neutral,
            text_layout: MessageTextLayout::Left,
        }
    }
}

impl FromNumber for MessageboxStyle {
    fn from_number(number: i32) -> Self {
        assert!(number >= 0);
        let msgbox_type = number & 0xf;
        let text_layout = (number >> 4) & 0xf;
        Self {
            messagebox_type: num_traits::FromPrimitive::from_i32(msgbox_type).unwrap_or_else(
                || panic!("MsgInit::from: unknown messagebox type: {}", msgbox_type),
            ),
            text_layout: num_traits::FromPrimitive::from_i32(text_layout)
                .unwrap_or_else(|| panic!("MsgInit::from: unknown text layout: {}", text_layout)),
        }
    }
}

/// A volume value, in the range [0.0, 1.0].
#[derive(Debug, Copy, Clone, PartialEq)]
pub struct Volume(pub f32);

impl Default for Volume {
    fn default() -> Self {
        Self(1.0)
    }
}

impl FromNumber for Volume {
    fn from_number(number: i32) -> Self {
        Self((number as f32 / 1000.0).clamp(0.0, 1.0)) // TODO: warn if out of range
    }
}

/// Defines a pan value in the range [-1.0, 1.0], where `0.0` is the center and `-1.0` is the hard left and `1.0` is the hard right.
#[derive(Debug, Copy, Clone, PartialEq)]
pub struct Pan(pub f32);

impl Default for Pan {
    fn default() -> Self {
        Self(0.0)
    }
}

impl FromNumber for Pan {
    fn from_number(number: i32) -> Self {
        Self((number as f32 / 1000.0).clamp(-1.0, 1.0)) // TODO: warn if out of range
    }
}
