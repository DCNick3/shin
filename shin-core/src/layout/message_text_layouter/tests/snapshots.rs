//! Tests based on single layout snapshots

use insta::assert_debug_snapshot;

use super::make_snapshot;
use crate::vm::command::types::{MessageTextLayout, MessageboxType};

#[test]
fn snapshot_ep1_nanjo1() {
    // here we have rubi text that is wider than the base text. the pointers have to be adjusted
    let snapshot = make_snapshot(MessageTextLayout::Justify, MessageboxType::Ushiromiya, "南條\u{3000}輝正@r@v19/11900001.「…………また。@k@v19/11900002.…お酒を@bたしな.@<嗜@>まれましたな？」");

    assert_debug_snapshot!("ep1_nanjo1", snapshot);
}

#[test]
fn snapshot_ep1_nanjo2() {
    // here we have a pretty long message with multiple Waits and spanning multiple lines
    let snapshot = make_snapshot(MessageTextLayout::Justify, MessageboxType::Ushiromiya, "南條\u{3000}輝正@r@v19/11900003.「…………金蔵さん。@k@v19/11900004.…あんたの体が一見調子がいいのは薬が効いてるからだ。@k@v19/11900005.だが、そんな強い酒を飲み続けては薬の意味もなくなってしまう。@k@v19/11900006.…悪いことは言わん。@k@v19/11900007.酒は控えなさい」");

    assert_debug_snapshot!("ep1_nanjo2", snapshot);
}

#[test]
fn snapshot_ep1_kinzo1() {
    // here we have rubi text that is smaller than the base text
    let snapshot = make_snapshot(MessageTextLayout::Justify, MessageboxType::Ushiromiya, "右代宮\u{3000}金蔵@r@v01/11500012.「はっはっはっは…。@k@v01/11500013.お前とて、指し間違えた手を待てと言っても聞かぬではないか。@k@v01/11500014.ならば@bあいこ.@<相子@>というものだろう」");

    assert_debug_snapshot!("ep1_kinzo1", snapshot);
}

#[test]
fn snapshot_ep1_kumasawa_nvl() {
    // tests novel mode, and, apparently, kinsoku shori-forbidden character on start of the line
    let snapshot = make_snapshot(MessageTextLayout::Justify, MessageboxType::Novel, "熊沢\u{3000}チヨ@r@v18/11800041_1.……おいたわしや、紗音さん、嘉音さん…。@k@r@v18/11800041_2.あの二人がいじめられる理由は何もないのです。@k@r@v18/11800041_3.…しかし、郷田さんに嫌われているのは紛れもない事実…。");

    assert_debug_snapshot!("ep1_kumasawa_nvl", snapshot);
}

#[test]
fn snapshot_ep2_beatrice_red_truth() {
    // tests red truth (instant text + color change), as well as section & sync commands, used to issue VM commands at certain points in the message
    let snapshot = make_snapshot(MessageTextLayout::Justify, MessageboxType::WitchSpace, "ベアトリーチェ@r@v27/20700938.「そうだ。@k@v27/20700939.だから妾はこれより、@v27/20700940.@|@y@c900.@[真実を語る時、赤を使うことにする@]@c.」");

    assert_debug_snapshot!("ep2_beatrice_red_truth", snapshot);
}

#[test]
fn snapshot_ep4_beatrice_blue_truth() {
    // tests 2 blue truths in one message
    let snapshot = make_snapshot(MessageTextLayout::Justify, MessageboxType::Novel, "ベアトリーチェ@r@v27/40700828.@|@y@c279.@[以上の復唱要求、並びに復唱拒否から、@]@k@v27/40700829.@|@y@[妾はそなたに対戦相手の資格がないことを宣言する。@]@c.");

    assert_debug_snapshot!("ep4_beatrice_blue_truth", snapshot);
}

#[test]
fn snapshot_ep1_novel_centered() {
    // test centered text in NVL mode
    let snapshot = make_snapshot(
        MessageTextLayout::Center,
        MessageboxType::Novel,
        "@r@a1666.ダカラ、@w666.現金ガスグニ大量ニ、@r@w666.喉カラ手ガ出ルホドニ欲シカッタ…！",
    );

    assert_debug_snapshot!("ep1_novel_centered", snapshot);
}

#[test]
fn snapshot_ep1_narrator1() {
    // this tests a non-novel message without a character name
    let snapshot = make_snapshot(
        MessageTextLayout::Justify,
        MessageboxType::Ushiromiya,
        "@r聴診器を外しながら、年輩の医師は溜め息を漏らす。",
    );

    assert_debug_snapshot!("ep1_narrator1", snapshot);
}

#[test]
fn snapshot_ep1_kinzo2() {
    // tests the correct implementation of multiline quoted text
    let snapshot = make_snapshot(
        MessageTextLayout::Justify,
        MessageboxType::Ushiromiya,
        "右代宮\u{3000}金蔵@r@v01/11500003.「忠告の気持ちだけはありがたくいただいておく。@k@v01/11500004.我が友よ。@k@v01/11500005.………源次。@k@v01/11500006.もう一杯頼む。@k@v01/11500007.心持ち薄めでな。@k@v01/11500008.南條の顔も立ててやれ」",
    );

    assert_debug_snapshot!("ep1_kinzo2", snapshot);
}

#[test]
fn snapshot_ep1_battler1() {
    // tests the correct handling of prohibition rules in combination with rubi text
    let snapshot = make_snapshot(
        MessageTextLayout::Justify,
        MessageboxType::Ushiromiya,
        "右代宮\u{3000}戦人@r@v10/10100014.「こんだけ@bた.@<立@>っ@bぱ.@<端@>がありゃー俺の発育は充分っすから～！@k\u{3000}@v10/10100015.むしろちょいと身長縮めた方が服が探しやすいくらい！」",
    );

    assert_debug_snapshot!("ep1_battler1", snapshot);
}

#[test]
fn snapshot_ep1_narrator2() {
    // tests the correctness of time reflowing when rubi and base texts have equal width
    let snapshot = make_snapshot(
        MessageTextLayout::Justify,
        MessageboxType::Ushiromiya,
        "@r@bねこぐるま.@<猫車@>のひとつしかない車輪が小石でも噛んだのでバランスを崩してしまったのだろう。",
    );

    assert_debug_snapshot!("ep1_narrator2", snapshot);
}

#[test]
fn snapshot_ep1_kinzo3() {
    // tests correctness of squish/justify selection when max_width = layout_width
    let snapshot = make_snapshot(
        MessageTextLayout::Justify,
        MessageboxType::Ushiromiya,
        "右代宮\u{3000}金蔵@r@v01/11500136.「留弗夫の間抜けは女遊びばかりッ！！@k\u{3000}@v01/11500137.楼座はどこの馬の骨ともわからん男の赤ん坊など生みおって！！@k\u{3000}@v01/11500138.朱志香は無能で無学だ！！@k\u{3000}@v01/11500139.譲治には男としての器がない！@k\u{3000}@v01/11500140.戦人は右代宮家の栄誉を自ら捨ておった愚か者だッ！！@k\u{3000}@v01/11500141.真里亞など見るのも汚らわしいッ！！」",
    );

    assert_debug_snapshot!("ep1_kinzo3", snapshot);
}
