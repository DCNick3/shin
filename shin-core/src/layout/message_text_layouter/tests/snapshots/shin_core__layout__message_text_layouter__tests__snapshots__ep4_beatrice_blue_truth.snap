---
source: shin-core/src/layout/message_text_layouter/tests/snapshots.rs
expression: snapshot
---
(
    [
        Voice(
            Voice {
                time: 0.0,
                line_index: 0,
                filename: "27/40700828",
                volume: 1.0,
                lipsync_enabled: true,
                time_to_first_sync: 0,
            },
        ),
        Section(
            Section {
                time: 0.0,
                line_index: 0,
                index: 1,
            },
        ),
        Sync(
            Sync {
                time: 0.0,
                line_index: 0,
                index: 0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: '以',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    0.0,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: '上',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    50.00002,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: 'の',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    100.00004,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: '復',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    150.00006,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: '唱',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    200.00008,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: '要',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    250.00009,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: '求',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: true,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    300.00012,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: '、',
                is_rubi: false,
                cant_be_at_line_start: true,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    350.00012,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: '並',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    400.00015,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: 'び',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    450.00015,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: 'に',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    500.00018,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: '復',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    550.00024,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: '唱',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    600.00024,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: '拒',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    650.0002,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: '否',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    700.0002,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: 'か',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    750.0002,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: 'ら',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: true,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    800.0002,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.001,
                line_index: 0,
                font: Regular,
                codepoint: '、',
                is_rubi: false,
                cant_be_at_line_start: true,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    850.0002,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Wait(
            Wait {
                time: 0.001,
                line_index: 0,
                is_last_wait: false,
                is_auto_click: false,
            },
        ),
        Voice(
            Voice {
                time: 0.002,
                line_index: 0,
                filename: "27/40700829",
                volume: 1.0,
                lipsync_enabled: true,
                time_to_first_sync: 0,
            },
        ),
        Section(
            Section {
                time: 0.002,
                line_index: 0,
                index: 2,
            },
        ),
        Sync(
            Sync {
                time: 0.002,
                line_index: 0,
                index: 1,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 0,
                font: Regular,
                codepoint: '妾',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    900.0001,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 0,
                font: Regular,
                codepoint: 'は',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    950.0001,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 0,
                font: Regular,
                codepoint: 'そ',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    1000.0001,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 0,
                font: Regular,
                codepoint: 'な',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    1050.0001,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 0,
                font: Regular,
                codepoint: 'た',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    1100.0001,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 0,
                font: Regular,
                codepoint: 'に',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    1150.0001,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 0,
                font: Regular,
                codepoint: '対',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    1200.0001,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 0,
                font: Regular,
                codepoint: '戦',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    1250.0001,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 0,
                font: Regular,
                codepoint: '相',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    1300.0,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 0,
                font: Regular,
                codepoint: '手',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    1350.0,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 0,
                font: Regular,
                codepoint: 'の',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    1400.0,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 0,
                font: Regular,
                codepoint: '資',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.00002,
                height: 50.0,
                position: Vec2(
                    1450.0,
                    60.625,
                ),
                horizontal_scale: 0.50505066,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 1,
                font: Regular,
                codepoint: '格',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    0.0,
                    134.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 1,
                font: Regular,
                codepoint: 'が',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    50.000122,
                    134.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 1,
                font: Regular,
                codepoint: 'な',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    100.000244,
                    134.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 1,
                font: Regular,
                codepoint: 'い',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    150.00037,
                    134.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 1,
                font: Regular,
                codepoint: 'こ',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    200.00049,
                    134.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 1,
                font: Regular,
                codepoint: 'と',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    250.00061,
                    134.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 1,
                font: Regular,
                codepoint: 'を',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    300.00073,
                    134.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 1,
                font: Regular,
                codepoint: '宣',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    350.00085,
                    134.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 1,
                font: Regular,
                codepoint: '言',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    400.00098,
                    134.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 1,
                font: Regular,
                codepoint: 'す',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    450.0011,
                    134.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 1,
                font: Regular,
                codepoint: 'る',
                is_rubi: false,
                cant_be_at_line_start: false,
                cant_be_at_line_end: true,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    500.00122,
                    134.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Char(
            Char {
                time: 0.003,
                line_index: 1,
                font: Regular,
                codepoint: '。',
                is_rubi: false,
                cant_be_at_line_start: true,
                cant_be_at_line_end: false,
                has_rubi: false,
                width: 50.000153,
                height: 50.0,
                position: Vec2(
                    550.00134,
                    134.625,
                ),
                horizontal_scale: 0.50505203,
                scale: 0.5208333,
                color: UnormColor(
                    4281976575,
                ),
                fade: 0.0,
            },
        ),
        Wait(
            Wait {
                time: 0.003,
                line_index: 1,
                is_last_wait: true,
                is_auto_click: false,
            },
        ),
    ],
    [
        LineInfo {
            width: 1500.0,
            y_position: 0.0,
            line_height: 70.0,
            baseline_ascent: 60.625,
            rubi_height: 20.0,
        },
        LineInfo {
            width: 600.00146,
            y_position: 74.0,
            line_height: 70.0,
            baseline_ascent: 60.625,
            rubi_height: 20.0,
        },
    ],
    Vec2(
        1500.0,
        144.0,
    ),
)
